forage-android/src/main/java/com/joinforage/forage/android/ForageSDK.kt

```diff
@@ -19,21 +20,70 @@ import com.joinforage.forage.android.network.data.CheckBalanceRepository
 import com.joinforage.forage.android.network.model.ForageApiResponse
 import com.joinforage.forage.android.network.model.ForageError
 import com.joinforage.forage.android.ui.ForagePINEditText
+import com.launchdarkly.sdk.LDContext
+import com.launchdarkly.sdk.android.LDClient
+import com.launchdarkly.sdk.android.LDConfig
+import com.launchdarkly.sdk.android.integrations.TestData
 import java.util.UUID
 
+internal object VaultConstants {
+    const val VGS_VAULT_TYPE = "vgs_vault_type"
+    const val BT_VAULT_TYPE = "bt_vault_type"
+}
+
+internal object LDConstants {
+    const val VAULT_TYPE_FLAG = "vault-primary-traffic-percentage"
+    const val USER = "anonymous-user"
+}
+
 /**
  * Singleton responsible for implementing the SDK API
  */
 object ForageSDK : ForageSDKApi {
     private var panEntry: PanEntry = PanEntry.Invalid("")
-    private val logger = Logger.getInstance(BuildConfig.DEBUG)
+    private var vaultType: String? = null
+    private const val LD_MOBILE_KEY = BuildConfig.LD_MOBILE_KEY
+
+    // vaultType is instantiated lazily and is a singleton. Once we set the vault type once, we don't
+    // want to overwrite it! We must take in the application as a parameter, which means that a
+    // ForagePINEditText must be rendered before any of the ForageSDKApi functions are called.
+    internal fun getVaultProvider(app: Application, dataSource: TestData? = null): String {
+        if (vaultType != null) {
+            return vaultType as String
+        }
+        // Datasource is required for testing purposes!
+        val ldConfig = if (dataSource != null) {
+            LDConfig.Builder()
+                .mobileKey(LD_MOBILE_KEY)
+                .dataSource(dataSource)
+                .build()
+        } else {
+            LDConfig.Builder()
+                .mobileKey(LD_MOBILE_KEY)
+                .build()
+        }
+        val context = LDContext.create(LDConstants.USER)
+        val client = LDClient.init(app, ldConfig, context, 0)
+        val vaultPercent = client.doubleVariation(LDConstants.VAULT_TYPE_FLAG, 0.0)
+        val randomNum = Math.random() * 100
+        vaultType = VaultConstants.BT_VAULT_TYPE
+        if (randomNum < vaultPercent) {
+            vaultType = VaultConstants.VGS_VAULT_TYPE
+        }
```

Comment:
shouldn't vaultType = VGS_VAULT_TYPE if we set the flag value to 0% (e.g. 100% vgs usage)?

we should also consider the case where we can't connect to LD and default to 0 (e.g. should we default to BT or VGS in the case of being unable to reach LD)?

I think we can safely re-use Shardendu's logic on the backend here:

```kotlin
val randomNum = (1..100).random().toDouble()
if (randomNum < vaultPercent) {
  vaultType = VaultConstants.BT_VAULT_TYPE
} else {
  vaultType = VaultConstants.VGS_VAULT_TYPE
}
```

Reply:
Thanks for the calllout, I flipped the logic here: [9f2e3dc](https://github.com/teamforage/forage-android-sdk/pull/31/commits/9f2e3dcfba1559e369e43ded24af68f69a99772d). The default is return BT if anything goes wrong.